# Лабораторная работа №2

**Table of contents**<a id='toc0_'></a>
- 1. [Описание задачи и признаков](#tocc1_)
    - 1.1. [Постановка задачи](#toc1_1_) 
    - 1.2. [Описание признакового пространства](#toc1_2_) 
    - 1.3. [Формировка виртуального окружения](#toc1_3_) 
        - 1.3.1. [Импорт библиотек](#toc1_3_1_) 
        - 1.3.2. [Версия Python](#toc1_3_2_)
        - 1.3.3. [Фиксация seed для воспроизводимости результатов](#toc1_3_3_)
- 2. [Предварительная обработка данных](#toc2_)
    - 2.1. [Загрузка данных](#toc2_1_)
    - 2.2. [Анализ данных](#toc2_2_)
    - 2.3. [Приведение типов данных и обработка некорректных значений](#toc2_3_)
        - 2.3.1. [Приведение пропусков к общему виду](#toc2_3_1_)
        - 2.3.2. [Кодирование категориальных данных](#toc2_3_2_)
    - 2.4. [Удаление неинформативных признаков](#toc2_4_)
- 3. [Разделение данных на обучающую и тестовую выборки](#toc3_)
    - 3.1. [Выделение целевого признака и предикторов](#toc3_1_)
    - 3.2. [Разбиение на обучающую и тестовую выборки](#toc3_2_)

## 1. <a id='tocc1_'></a>[Описание задачи и признаков](#toc0_)

### 1.1. <a id='toc1_1_'></a>[Постановка задачи](#toc0_)

**Задача:** Прогнозирование коэффициента `Гармония Бессмертия` — ключевого показателя состояния магического портала "Врата Вечности".

**Тип задачи:** Задача **регрессии**, так как целевая переменная (`Гармония Бессмертия`) является непрерывной числовой величиной (коэффициент в диапазоне от 0 до 1).

**Цель:** Построить модель машинного обучения, которая по параметрам работы портала сможет предсказать значение его Гармонии. Это позволит заблаговременно обнаружить критический износ и предотвратить разрушение портала.

**Целевая переменная:**
- `Гармония Бессмертия`

**Входные данные:** 19 признаков, описывающих различные магические и физические параметры работы портала.

### 1.2. <a id='toc1_2_'></a>[Описание признакового пространства](#toc0_)

| **№** | Признак | Описание | Тип |
|---|---|---|---|
| | **Целевые признаки** | | |
| **1** | `Гармония Бессмертия` | Коэффициент, указывающий на состояние магического ядра портала и его стабильность | Вещественный(0-1) |
| | **Входные признаки** | | |
| **2** | `Вектор Мощи` | Позиция магического механизма, управляющего интенсивностью перехода через портал | Вещественный |
| **3** | `Скорость перехода через портал` | Скорость, с которой портал перемещает объект через пространство | Вещественный |
| **4** | `Приток Силы Потока` | Поток магической силы, текущий через ядро портала, обеспечивая стабильную работу | Вещественный |
| **5** | `Ритм магического ядра` | Число оборотов магического ядра портала в минуту | Вещественный |
| **6** | `Поток Энергий` | Число оборотов генератора эфира, преобразующего внешнюю энергию в магическую | Вещественный |
| **7** | `Сила Левого Потока` | Мощность магического потока, текущего через левую сторону портала | Вещественный |
| **8** | `Сила Правого Потока` | Мощность магического потока, текущего через правую сторону портала | Вещественный |
| **9** | `Пламя Стихий` | Температура магической энергии, исходящей из высокого магического источника портала (°C) | Вещественный |
| **10** | `Температура вдоха Истока` | Температура воздуха, входящего в магический ускоритель портала (°C) | Целочисленный |
| **11** | `Температура выдоха Истока` | Температура воздуха, исходящего из магического ускорителя портала (°C) | Вещественный |
| **12** | `Приток давления Выдоха Истока` | Давление магического потока на выходе из магического источника высокого давления | Вещественный |
| **13** | `Давление вдоха Истока` | Давление воздуха, входящего в магический ускоритель | Вещественный |
| **14** | `Давление выдоха Истока` | Давление воздуха, исходящего из магического ускорителя | Вещественный |
| **15** | `Древний Ветер` | Давление древней магической энергии, покидающей портал в виде выхлопного потока | Вещественный |
| **16** | `Печать Чародея` | Параметр управления магическим выбросом в сердце портала (%) | Вещественный |
| **17** | `Эмульсия Истока` | Количество магического топлива, подпитывающего портал | Вещественный |
| **18** | `Дыхание Истока` | Коэффициент, отображающий степень угасания магического ускорителя | Вещественный |
| **19** | `Тип Вектора Мощи` | Метка типа позиции магического механизма | Категориальный |
| **20** | `Номер пометки` | Уникальный номер записи о работе портала | Целочисленный |

### 1.3. <a id='toc1_3_'></a>[Формировка виртуального окружения](#toc0_)

#### 1.3.1. <a id='toc1_3_1_'></a>[Импорт библиотек](#toc0_)

In [22]:
import sys

import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings('ignore')

#### 1.3.2. <a id='toc1_3_2_'></a>[Версия Python](#toc0_)

In [23]:
print(f"Python: {sys.version}")

Python: 3.11.9 (tags/v3.11.9:de54cf5, Apr  2 2024, 10:12:12) [MSC v.1938 64 bit (AMD64)]


#### 1.3.3. <a id='toc1_3_3_'></a>[Фиксация seed для воспроизводимости результатов](#toc0_)

In [24]:
seed = 42
np.random.seed(seed)

## 2. <a id='toc2_'></a>[Предварительная обработка данных](#toc0_)

### 2.1. <a id='toc2_1_'></a>[Загрузка данных](#toc0_)

In [25]:
df = pd.read_csv('portal_data.csv', encoding='windows-1251', sep='|')

Вывод первых строк для проверки:

In [26]:
df.head()

,Вектор Мощи,Скорость перехода через портал,Приток Силы Потока,Ритм магического ядра,Поток Энергий,Сила Левого Потока,Сила Правого Потока,Пламя Стихий,Температура вдоха Истока,Температура выдоха Истока,Приток давления Выдоха Истока,Давление вдоха Истока,Давление выдоха Истока,Древний Ветер,Печать Чародея,Эмульсия Истока,Дыхание Истока,Гармония Бессмертия,Тип Вектора Мощи,Номер пометки
0,1.138,1.54332,289.964,141.318158,6677.380,7.584,7.584,464.006,736,-,1.096,Не определено,5.947,-,7.137,0.082,0.95,0.975,Слабый,0
1,2.088,-,6960.180,144.111767,6828.469,28.204,28.204,635.401,736,581.658,1.331,1.394603,7.282,1.019,10.655,0.287,0.95,0.975,Слабый,1
2,3.144,4.6299600000000005,8379.229,145.220853,7111.811,60.358,60.358,606.002,736,-,1.389,Не определено,7.574,-,13.086,0.259,0.95,0.975,Ниже среднего,2
3,4.161,6.17328,14724.395,162.050156,7792.630,113.774,113.774,661.471,736,-,1.658,Не определено,9.007,-,18.109,0.358,0.95,0.975,Ниже среднего,3
4,5.140,7.7166,21636.432,201.513586,8494.777,175.306,175.306,731.494,736,645.642,2.078,Не определено,11.197,1.026,26.373,0.522,0.95,0.975,Выше среднего,4


### 2.2. <a id='toc2_2_'></a>[Анализ данных](#toc0_)

Общая информация о датасете:

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11934 entries, 0 to 11933
Data columns (total 20 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Вектор Мощи                     11934 non-null  float64
 1   Скорость перехода через портал  11934 non-null  object 
 2   Приток Силы Потока              11934 non-null  float64
 3   Ритм магического ядра           11934 non-null  float64
 4   Поток Энергий                   11934 non-null  float64
 5   Сила Левого Потока              11934 non-null  float64
 6   Сила Правого Потока             11934 non-null  float64
 7   Пламя Стихий                    11934 non-null  float64
 8   Температура вдоха Истока        11934 non-null  int64  
 9   Температура выдоха Истока       11934 non-null  object 
 10  Приток давления Выдоха Истока   11934 non-null  float64
 11  Давление вдоха Истока           11934 non-null  object 
 12  Давление выдоха Истока          

In [28]:
df.describe()

,Вектор Мощи,Приток Силы Потока,Ритм магического ядра,Поток Энергий,Сила Левого Потока,Сила Правого Потока,Пламя Стихий,Температура вдоха Истока,Приток давления Выдоха Истока,Давление выдоха Истока,Печать Чародея,Эмульсия Истока,Дыхание Истока,Гармония Бессмертия,Номер пометки
count,11934.000000,11934.000000,11934.000000,11934.000000,11934.000000,11934.000000,11934.000000,11934.0,11934.000000,11934.000000,11934.000000,11934.000000,11934.00000,11934.0000,11934.000000
mean,5.166667,27247.498685,223.711688,8200.947312,227.335768,227.335768,735.495446,736.0,2.352963,12.297123,33.641261,0.662440,0.97500,0.9875,5966.500000
std,2.626388,22148.613155,81.061874,1091.315507,200.495889,200.495889,173.680552,0.0,1.084770,5.337448,25.841363,0.507132,0.01472,0.0075,3445.193391
min,1.138000,253.547000,136.939406,6589.002000,5.304000,5.304000,442.364000,736.0,1.093000,5.828000,0.000000,0.068000,0.95000,0.9750,0.000000
25%,3.144000,8375.883750,145.220958,7058.324000,60.317000,60.317000,589.872750,736.0,1.389000,7.447250,13.677500,0.246000,0.96200,0.9810,2983.250000
50%,5.140000,21630.659000,201.514947,8482.081500,175.268000,175.268000,706.038000,736.0,2.083000,11.092000,25.276500,0.496000,0.97500,0.9875,5966.500000
75%,7.148000,39001.426750,280.447777,9132.606000,332.364750,332.364750,834.066250,736.0,2.981000,15.658000,44.552500,0.882000,0.98800,0.9940,8949.750000
max,9.300000,72784.872000,372.879926,9797.103000,645.249000,645.249000,1115.797000,736.0,4.560000,23.140000,92.556000,1.832000,1.00000,1.0000,11933.000000


Рассмотрим датасет на содержание уникальных значений:

In [29]:
df.nunique()

Вектор Мощи                           9
Скорость перехода через портал       10
Приток Силы Потока                11430
Ритм магического ядра              3888
Поток Энергий                     11834
Сила Левого Потока                 4286
Сила Правого Потока                4286
Пламя Стихий                      11772
Температура вдоха Истока              1
Температура выдоха Истока          2369
Приток давления Выдоха Истока       524
Давление вдоха Истока                 2
Давление выдоха Истока             4209
Древний Ветер                        20
Печать Чародея                     8496
Эмульсия Истока                     696
Дыхание Истока                       51
Гармония Бессмертия                  26
Тип Вектора Мощи                      4
Номер пометки                     11934
dtype: int64

Анализ категориального признака:

In [30]:
categorical_column = 'Тип Вектора Мощи'
print(f"Категориальный признак '{categorical_column}':")
print(f"Количество уникальных значений: {df[categorical_column].nunique()}")
print(f"Уникальные значения: {df[categorical_column].unique()}")
print(f"Распределение значений:")
print(df[categorical_column].value_counts())

Категориальный признак 'Тип Вектора Мощи':
Количество уникальных значений: 4
Уникальные значения: ['Слабый' 'Ниже среднего' 'Выше среднего' 'Сильный']
Распределение значений:
Тип Вектора Мощи
Сильный          3978
Слабый           2652
Ниже среднего    2652
Выше среднего    2652
Name: count, dtype: int64


Сравнение индентичных на первый взгляд признаков:

In [31]:
mismatch_count = (df['Сила Левого Потока'] != df['Сила Правого Потока']).sum()
print(f"Количество несовпадений: {mismatch_count}")

Количество несовпадений: 0


**Проблемы, выявленные при анализе:**
- Несколько столбцов имеют тип object вместо числового типа из-за наличия строковых значений ("-", "Не определено")
- Признак `Температура вдоха Истока` имеет только 1 уникальное значение
- Признак `Номер пометки` имеет 11934 уникальных значения (идентификатор)
- `Сила Левого Потока` и `Сила Правого Потока` имеют абсолютно одинаковые значения
- Аномальное значение в столбце `Скорость перехода через портал`(4.6299600000000005)

### 2.3. <a id='toc2_3_'></a>[Приведение типов данных и обработка некорректных значений](#toc0_)

Округляем до 5 знаков после запятой (разумная точность для физического параметра)

In [32]:
df['Скорость перехода через портал'] = df['Скорость перехода через портал'].round(5)

#### 2.3.1. <a id='toc2_3_1_'></a>[Приведение пропусков к общему виду](#toc0_)

Поскольку несколько столбцов имеют тип object из-за наличия строковых значений "-" и "Не определено". Преобразуем их в числовой формат:

In [33]:
object_columns = [
    'Скорость перехода через портал',
    'Температура выдоха Истока', 
    'Давление вдоха Истока',
    'Древний Ветер'
]

for col in object_columns:
    df[col] = pd.to_numeric(df[col].replace(['-', 'Не определено'], np.nan), errors='coerce')

print("Типы данных после преобразования:")
print(df[object_columns].dtypes)
print(f"\nКоличество пропусков после преобразования:")
print(df[object_columns].isnull().sum())

Типы данных после преобразования:
Скорость перехода через портал    float64
Температура выдоха Истока         float64
Давление вдоха Истока             float64
Древний Ветер                     float64
dtype: object

Количество пропусков после преобразования:
Скорость перехода через портал    1203
Температура выдоха Истока         9547
Давление вдоха Истока             9495
Древний Ветер                     9547
dtype: int64


Теперь можно рассмотреть количество пропусков в каждом признаке

In [34]:
(df.isna().sum()/df.shape[0]*100).sort_values(ascending=False)

Древний Ветер                     79.998324
Температура выдоха Истока         79.998324
Давление вдоха Истока             79.562594
Скорость перехода через портал    10.080442
Ритм магического ядра              0.000000
Приток Силы Потока                 0.000000
Вектор Мощи                        0.000000
Поток Энергий                      0.000000
Пламя Стихий                       0.000000
Сила Правого Потока                0.000000
Температура вдоха Истока           0.000000
Сила Левого Потока                 0.000000
Приток давления Выдоха Истока      0.000000
Давление выдоха Истока             0.000000
Печать Чародея                     0.000000
Эмульсия Истока                    0.000000
Дыхание Истока                     0.000000
Гармония Бессмертия                0.000000
Тип Вектора Мощи                   0.000000
Номер пометки                      0.000000
dtype: float64

#### 2.3.2. <a id='toc2_3_2_'></a>[Кодирование категориальных данных](#toc0_)

In [35]:
categorical_column = 'Тип Вектора Мощи'

print("Категориальный признак до кодирования:")
print(f"Уникальные значения: {df[categorical_column].unique()}")
print(f"Распределение значений:")
print(df[categorical_column].value_counts())

order_mapping = {
    'Слабый': 0,
    'Ниже среднего': 1, 
    'Выше среднего': 2,
    'Сильный': 3
}

df[categorical_column] = df[categorical_column].map(order_mapping)

print("\nКатегориальный признак после кодирования:")
print(f"Уникальные значения: {df[categorical_column].unique()}")
print(f"Распределение значений:")
print(df[categorical_column].value_counts().sort_index())
print(f"Тип данных: {df[categorical_column].dtype}")

Категориальный признак до кодирования:
Уникальные значения: ['Слабый' 'Ниже среднего' 'Выше среднего' 'Сильный']
Распределение значений:
Тип Вектора Мощи
Сильный          3978
Слабый           2652
Ниже среднего    2652
Выше среднего    2652
Name: count, dtype: int64

Категориальный признак после кодирования:
Уникальные значения: [0 1 2 3]
Распределение значений:
Тип Вектора Мощи
0    2652
1    2652
2    2652
3    3978
Name: count, dtype: int64
Тип данных: int64


### 2.4. <a id='toc2_4_'></a>[Удаление неинформативных признаков](#toc0_)

Удаляем признаки, которые не несут информационной ценности для модели
1. `Номер пометки` - уникальный идентификатор
2. `Температура вдоха Истока` - имеет только 1 уникальное значение
3. `Сила Правого Потока` - полностью дублирует `Сила Левого Потока`
4. `Древний Ветер`, `Температура выдоха Истока`, `Давление вдоха Истока` имеют около 80% пропущенных значений

In [36]:
columns_to_drop = ['Номер пометки', 'Температура вдоха Истока', 'Сила Правого Потока', 'Древний Ветер', 'Температура выдоха Истока', 'Давление вдоха Истока']
df = df.drop(columns=columns_to_drop)

Данные после обработки:

In [37]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11934 entries, 0 to 11933
Data columns (total 14 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Вектор Мощи                     11934 non-null  float64
 1   Скорость перехода через портал  10731 non-null  float64
 2   Приток Силы Потока              11934 non-null  float64
 3   Ритм магического ядра           11934 non-null  float64
 4   Поток Энергий                   11934 non-null  float64
 5   Сила Левого Потока              11934 non-null  float64
 6   Пламя Стихий                    11934 non-null  float64
 7   Приток давления Выдоха Истока   11934 non-null  float64
 8   Давление выдоха Истока          11934 non-null  float64
 9   Печать Чародея                  11934 non-null  float64
 10  Эмульсия Истока                 11934 non-null  float64
 11  Дыхание Истока                  11934 non-null  float64
 12  Гармония Бессмертия             

In [38]:
df.head(10)

,Вектор Мощи,Скорость перехода через портал,Приток Силы Потока,Ритм магического ядра,Поток Энергий,Сила Левого Потока,Пламя Стихий,Приток давления Выдоха Истока,Давление выдоха Истока,Печать Чародея,Эмульсия Истока,Дыхание Истока,Гармония Бессмертия,Тип Вектора Мощи
0,1.138,1.54332,289.964,141.318158,6677.380,7.584,464.006,1.096,5.947,7.137,0.082,0.95,0.975,0
1,2.088,NaN,6960.180,144.111767,6828.469,28.204,635.401,1.331,7.282,10.655,0.287,0.95,0.975,0
2,3.144,4.62996,8379.229,145.220853,7111.811,60.358,606.002,1.389,7.574,13.086,0.259,0.95,0.975,1
3,4.161,6.17328,14724.395,162.050156,7792.630,113.774,661.471,1.658,9.007,18.109,0.358,0.95,0.975,1
4,5.140,7.71660,21636.432,201.513586,8494.777,175.306,731.494,2.078,11.197,26.373,0.522,0.95,0.975,2
5,6.175,9.25992,29792.731,241.630782,8828.360,246.278,800.434,2.501,13.356,35.760,0.708,0.95,0.975,2
6,7.148,10.80324,38982.180,280.448510,9132.429,332.077,854.747,2.963,15.679,45.881,0.908,0.95,0.975,3
7,8.206,12.34656,50996.808,323.328632,9318.562,437.989,952.122,3.576,18.632,62.440,1.236,0.95,0.975,3
8,9.300,13.88988,72763.329,372.843693,9778.528,644.905,1115.797,4.498,22.811,92.556,1.832,0.95,0.975,3
9,1.138,1.54332,379.880,141.934538,6683.916,7.915,464.017,1.100,5.963,3.879,0.079,0.95,0.976,0


## 3. <a id='toc3_'></a>[Разделение данных на обучающую и тестовую выборки](#toc0_)

### 3.1. <a id='toc3_1_'></a>[Выделение целевого признака и предикторов](#toc0_)

In [39]:
target_column = 'Гармония Бессмертия'
X = df.drop(columns=[target_column])
y = df[target_column]

Объект признак:

In [40]:
print("матрица объект-признак:")
X

матрица объект-признак:


,Вектор Мощи,Скорость перехода через портал,Приток Силы Потока,Ритм магического ядра,Поток Энергий,Сила Левого Потока,Пламя Стихий,Приток давления Выдоха Истока,Давление выдоха Истока,Печать Чародея,Эмульсия Истока,Дыхание Истока,Тип Вектора Мощи
0,1.138,1.54332,289.964,141.318158,6677.380,7.584,464.006,1.096,5.947,7.137,0.082,0.95,0
1,2.088,NaN,6960.180,144.111767,6828.469,28.204,635.401,1.331,7.282,10.655,0.287,0.95,0
2,3.144,4.62996,8379.229,145.220853,7111.811,60.358,606.002,1.389,7.574,13.086,0.259,0.95,1
3,4.161,6.17328,14724.395,162.050156,7792.630,113.774,661.471,1.658,9.007,18.109,0.358,0.95,1
4,5.140,7.71660,21636.432,201.513586,8494.777,175.306,731.494,2.078,11.197,26.373,0.522,0.95,2
...,...,...,...,...,...,...,...,...,...,...,...,...,...
11929,5.140,7.71660,21624.934,201.516623,8470.013,175.239,681.658,2.087,10.990,23.803,0.471,1.00,2
11930,6.175,9.25992,29763.213,241.561772,8800.352,245.954,747.405,2.512,13.109,32.671,0.647,1.00,2
11931,7.148,10.80324,39003.867,280.444950,9120.889,332.389,796.457,2.982,15.420,42.104,0.834,1.00,3
11932,8.206,12.34656,50992.579,323.315332,9300.274,438.024,892.945,3.594,18.293,58.064,1.149,1.00,3


Целевые значения:

In [41]:
print("список целевых значений:")
y

список целевых значений:


0        0.975
1        0.975
2        0.975
3        0.975
4        0.975
         ...  
11929    1.000
11930    1.000
11931    1.000
11932    1.000
11933    1.000
Name: Гармония Бессмертия, Length: 11934, dtype: float64

### 3.2. <a id='toc3_2_'></a>[Разбиение на обучающую и тестовую выборки](#toc0_)

In [42]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

NameError: name 'train_test_split' is not defined

80/20 - стандартное соотношение, обеспечивающее достаточное количество данных для обучения

In [ ]:
print(f"\nРазмеры выборок:")
print(f"Обучающая: {X_train.shape}, {y_train.shape}")
print(f"Тестовая: {X_test.shape}, {y_test.shape}")


Размеры выборок:
Обучающая: (9547, 13), (9547,)
Тестовая: (2387, 13), (2387,)
